samplomatic/transpiler/passes/add_terminal_right_dressed_boxes.py

```diff
@@ -14,145 +14,180 @@
+             * Those qubits that should become active, if not already.
+             * Those qubits that the given operation presently terminates, without intervention.
         """
-        if not node.op.name == "box":
-            return uncollected_qubits
-
-        twirl = get_annotation(node.op, Twirl)
-        change_basis = get_annotation(node.op, ChangeBasis)
-        if twirl and twirl.dressing == "right":
-            # Right-dressed boxes act as collectors
-            uncollected_qubits = uncollected_qubits.difference(node.qargs)
-        elif twirl or change_basis:
-            uncollected_qubits = uncollected_qubits.union(node.qargs)
-            for sub_node in asap_topological_nodes(circuit_to_dag(node.op.body)):
-                if sub_node.op.name == "measure":
-                    # Measurements in a left-dressed box act as collectors
-                    uncollected_qubits = uncollected_qubits.difference(sub_node.qargs)
-        return uncollected_qubits
+        if node.is_standard_gate() and node.op.num_qubits == 1:
+            # single qubit gates outside of a box are re-applied without further consideration
+            return _EMPTY_SET, _EMPTY_SET, _EMPTY_SET
+
+        elif node.op.name == "box":
+            # boxes need special treatment depending on annotations
+            if (twirl := get_annotation(node.op, Twirl)) and twirl.dressing == "right":
+                # right-dressed boxes terminate uncollected qubits unconditionally
+                return _EMPTY_SET, _EMPTY_SET, set(node.qargs)
+
+            elif twirl or get_annotation(node.op, ChangeBasis):
+                # left-dressed boxes terminate uncollected qubits on all qubits they measure
+                # internally (twirled measurements), but they create/continue uncollected qubits
+                # elsewhere. first, traverse the box and see which qubits are measured.
+                measured_qubits = set()
+                qubit_map = None  # don't compute this unless necessary
+                for instr in node.op.body:
+                    if instr.operation.name == "measure":
+                        qubit_map = qubit_map or (qubit_map := _inverse_box_qubit_map(node))
+                        measured_qubits.update(qubit_map[qubit] for qubit in instr.qubits)
+
+                return _EMPTY_SET, set(node.qargs).difference(measured_qubits), measured_qubits
+            else:
+                # an un-annotated box might contain stuff that we can't propagate through, so
+                # it is safest to just terminate now. ambitious readers of this code could update
+                # it to be less lazy and traverse the contents to find out.
+                return set(node.qargs), _EMPTY_SET, _EMPTY_SET
+        elif node.op.name == "barrier":
+            # it's always okay to postpone termination until after a barrier
+            return _EMPTY_SET, _EMPTY_SET, _EMPTY_SET
+        else:
```

Comment:
Nit pick: I know that we have some validation such that we only expect some types of operations, and therefore here we'd expect to get a two-qubit gate. Yet, I have a preference for being more verbose in these kinds of situations, for:
- readability (it's explicit, the `else` is more ambiguous)
- if in the future we start supporting something else but we forget to update this pass, it'll error rather than do its own thing

```suggestion python
        elif node.is_standard_gate() and node.op.num_qubits == 2:
            ...
        else:
            raise
```

Reply:
okay, looks good